projects/driver_controls/src/ads1015.c

```diff
@@ -0,0 +1,151 @@
+#include <ads1015.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#include "ads1015_bitmasks.h"
+#include "delay.h"
+#include "gpio_it.h"
+#include "i2c.h"
+
+// Union definition to help process registers data
+typedef struct ADS1015Register {
+  uint8_t data[2];
+} ADS1015Register;
+
+// Since the ADS1015 stores data in two's complement, an int16_t will be used to hold
+// the raw readings
+typedef struct ADS1015Interrupt {
+  ADS1015Callback callback;
+  void *context;
+  int16_t reading;
+} ADS1015Interrupt;
+
+static I2CPort s_i2c_port;
+static ADS1015Channel s_current_channel;
+
+static ADS1015Interrupt s_interrupts[NUM_ADS1015_CHANNELS];
+
+static StatusCode prv_channel_valid(ADS1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  return STATUS_CODE_OK;
+}
+
+// Initiates a read as outlined in section 8.5.3 of the datasheet
+static StatusCode prv_read(I2CPort i2c_port, uint8_t reg, uint8_t *data, size_t len) {
+  status_ok_or_return(i2c_write(i2c_port, ADS1015_I2C_ADDRESS, &reg, 1));
```

Comment:
Magic 1? I thought reg was 2 bytes?

Reply:
The address pointer register is one byte,  and we need to write to it before every transaction so that the adc knows which register we want.